### PR TITLE
linux-rpi: 6.1.63-stable_20231123 -> 6.6.31-stable_20240529;  raspberrypifw: stable_20231123 -> 1.20240529;  raspberrypiWirelessFirmware: unstable-2023-11-15 -> unstable-2024-02-26

### PR DESCRIPTION
**Pull request by peat-psuwit**
## Description of changes

- linux-rpi is upgrade from 6.1.63 to 6.6.31. [This forum thread](https://forums.raspberrypi.com/viewtopic.php?t=361116) discuses the jump from 6.1.y to 6.6.y. Additional upstream 6.6.y versions and downstream patches are added since then.
- raspberrypi-firmware are binary blobs which are updated together with linux-rpi. [The latest commit](https://github.com/raspberrypi/firmware/compare/524247ac6d8b1f4ddd53730e978a70c76a320bd6...1.20240529) doesn't seem to do anything more that updating prebuilt kernels, but included here for completeness.
- [bluez-firmware](https://github.com/RPi-Distro/bluez-firmware/compare/d9d4741caba7314d6500f588b1eaa5ab387a4ff5...78d6a07730e2d20c035899521ab67726dc028e1c) and [firmware-nonfree](https://github.com/RPi-Distro/firmware-nonfree/compare/88aa085bfa1a4650e1ccd88896f8343c22a24055...223ccf3a3ddb11b3ea829749fbbba4d65b380897) are updated so that they use a "generic" firmware so that WPA3-SAE works - not sure what that means. Firmware-nonfree also see Debian packaging merge & additional firmware, however since we don't use those we can safely ignore them.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux (cross-compiled, I promised I tested both linux-rpi3 and linux-rpi4 this time)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lopsided98** [comment]
I've had some failures building linux_rpi for 32-bit lately, I wonder whether this fixes it.

@ofborg build pkgsCross.raspberryPi.linux_rpi0

**Comment by Cryolitia** [comment]
could you please update it to [stable_20240529](https://github.com/raspberrypi/linux/releases/tag/stable_20240529)

**Comment by peat-psuwit** [comment]
@Cryolitia linux_rpi is updated to [stable_20240529](https://github.com/raspberrypi/linux/releases/tag/stable_20240529), and raspberrypi-fw is updated to the corresponding 1.20240529. raspberrypiWirelessFirmware has no update since the last time I submitted this PR.

File: pkgs/os-specific/linux/kernel/linux-rpi.nix [hunk 1]
```diff
@@ -24,32 +24,31 @@ lib.overrideDerivation (buildLinux (args // {
...
   } // (args.features or {});
 
   kernelPatches = (args.kernelPatches or []) ++ [
-    # Fix "WARNING: unmet direct dependencies detected for MFD_RP1", and
-    # subsequent build failure.
-    # https://github.com/NixOS/nixpkgs/pull/268280#issuecomment-1911839809
-    # https://github.com/raspberrypi/linux/pull/5900
     {
-      name = "drm-rp1-depends-on-instead-of-select-MFD_RP1.patch";
+      name = "gpio-pwm_-_pwm_apply_might_sleep.patch";
```
Comment (Cryolitia) [inline pkgs/os-specific/linux/kernel/linux-rpi.nix:40]: could you please add more comment for patches

Comment (peat-psuwit) [reply pkgs/os-specific/linux/kernel/linux-rpi.nix:42]: Done. Please take another look.

Comment (Cryolitia) [reply pkgs/os-specific/linux/kernel/linux-rpi.nix:42]: > Done. Please take another look.

Also all other patchs pls.

**Review by Cryolitia — APPROVED** [review]
Successfully build and boot on my rpi5b, not sure how to further testing.

both tested aarch64 native compile and x86_64 cross compile 

LGTM so far

File: pkgs/os-specific/linux/kernel/linux-rpi.nix (hunk 1 above)
Comment (peat-psuwit) [reply pkgs/os-specific/linux/kernel/linux-rpi.nix:42]: Both 2 patches come together and is part of the same PR. Should I explicitly say so in the comment?

**Comment by Cryolitia** [comment]
could you please check the 「Fits CONTRIBUTING.md.」 in 「Things Done」 in your description lol

File: pkgs/os-specific/linux/kernel/linux-rpi.nix (hunk 1 above)
Comment (Cryolitia) [reply pkgs/os-specific/linux/kernel/linux-rpi.nix:42]: got it. that's enough, not need to explicitly comment,thx

File: pkgs/os-specific/linux/kernel/linux-rpi.nix [hunk 2]
```diff
@@ -24,32 +24,33 @@ lib.overrideDerivation (buildLinux (args // {
...
     efiBootStub = false;
   } // (args.features or {});
 
   kernelPatches = (args.kernelPatches or []) ++ [
-    # Fix "WARNING: unmet direct dependencies detected for MFD_RP1", and
-    # subsequent build failure.
-    # https://github.com/NixOS/nixpkgs/pull/268280#issuecomment-1911839809
-    # https://github.com/raspberrypi/linux/pull/5900
+    # Fix compilation errors due to incomplete patch backport.
+    # https://github.com/raspberrypi/linux/pull/6223
```
Comment (Cryolitia) [inline pkgs/os-specific/linux/kernel/linux-rpi.nix:40]: According to https://github.com/raspberrypi/linux/pull/6223#issuecomment-2173323194 Do we still need it now?

Comment (peat-psuwit) [reply pkgs/os-specific/linux/kernel/linux-rpi.nix:40]: The fixing commit is not yet part of `stable_20240529` tag.

Comment (Cryolitia) [reply pkgs/os-specific/linux/kernel/linux-rpi.nix:40]: thx for clarify

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**